traits/src/grid.rs

```diff
@@ -1,41 +1,89 @@
+use crate::cell::ReferenceCellType;
+use solvers_tools::arrays::AdjacencyList;
+use std::ops::Range;
 
 pub trait Geometry {
-    fn reference_cell(&self) -> &dyn ReferenceCell;
-    fn map(&self, reference_coords: &[f64], physical_coords: &mut [f64]);
+    //! Grid geometry
+    //!
+    //! Grid geometry provides information about the physical locations of mesh points in space
 
+    /// The geometric dimension
     fn dim(&self) -> usize;
-}
 
-pub enum Locality {
-    Local,
-    Ghost(usize),
-    Remote,
+    /// Get the point with the index `i`
+    fn point(&self, i: usize) -> Option<&[f64]>;
+
+    /// The number of points stored in the geometry
+    fn point_count(&self) -> usize;
+
+    /// Get the vertex numbers of a cell
+    fn cell_vertices(&self, index: usize) -> Option<&[usize]>;
 }
 
 pub trait Topology {
-    // The dimension of the grid
+    //! Grid topology
+    //!
+    //! Grid topology provides information about which mesh entities are connected to other mesh entities
+
+    /// The dimension of the grid
     fn dim(&self) -> usize;
 
-    // Check the locality of an element
-    fn locality(global_id: usize) -> Locality;
+    /// Return the index map from the input order to the storage order
+    fn index_map(&self) -> &[usize];
+
+    /// Get the indices of cells with the given cell type
+    fn get_cells(&self, cell_type: ReferenceCellType) -> Vec<usize>;
+
+    /// Get the indices of cells with the given cell type as a range (if they are contiguous
+    fn get_cells_range(&self, cell_type: ReferenceCellType) -> Option<Range<usize>>;
+
+    /// Convert local id of a cell to global id of the cell
+    fn local2global(&self, local_id: usize) -> usize;
 
-    // Convert local to global id
-    fn local2global(local_id: usize) -> usize;
+    /// Convert global id of a cell to local id of the cell
+    fn global2local(&self, global_id: usize) -> Option<usize>;
 
-    // Convert global to local id
-    fn global2local(global_id: usize) -> Option<usize>;
+    /// The number of entities of dimension `dim`
+    fn entity_count(&self, dim: usize) -> usize;
+
+    /// The indices of the vertices that from cell with index `index`
+    fn cell(&self, index: usize) -> Option<&[usize]>;
+
+    /// Create the connectivity of entities of dimension `dim0` to entities of dimension `dim1`
+    ///
+    /// If this function is called multiple times, it will do nothing after the first call
+    fn create_connectivity(&mut self, dim0: usize, dim1: usize);
+
+    /// Create the connectivity information for all dimensions
+    fn create_connectivity_all(&mut self) {
+        for dim0 in 0..self.dim() {
+            for dim1 in 0..self.dim() {
+                self.create_connectivity(dim0, dim1);
+            }
+        }
+    }
+
+    /// Get the connectivity of entities of dimension `dim0` to entities of dimension `dim1`
+    fn connectivity(&self, dim0: usize, dim1: usize) -> &AdjacencyList<usize>;
 }
```

Comment:
I've done this in https://github.com/rusty-fast-solvers/solvers/pull/45/commits/461f04289a113a69cb5db52607e90e23edc81894 but it means that grid have to be mutable in a lot of places now. Maybe there's a better way to do it

Reply:
IMO, this (an index or cache) is a classic use case for interior mutability (`RefCell` or `RwLock`). I think this can be revisited another time. It is possible to avoid an explicit representation and instead construct `cell->vertex` connectivity on the fly by traversal, which keeps the memory smaller. I feel like there's usually one specific kind of traversal that is perf-sensitive and everything else would ideally be done on the fly.